artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/proton/handler/ProtonHandler.java

```diff
@@ -322,17 +308,22 @@ private void dispatch() {
       // because we could have a distributed deadlock
       // while processing events (for instance onTransport)
       // while a client is also trying to write here
-      while ((ev = popEvent()) != null) {
-         for (EventHandler h : handlers) {
-            if (log.isTraceEnabled()) {
-               log.trace("Handling " + ev + " towards " + h);
-            }
-            try {
-               Events.dispatch(ev, h);
-            } catch (Exception e) {
-               log.warn(e.getMessage(), e);
-               connection.setCondition(new ErrorCondition());
+
+      synchronized (lock) {
+         while ((ev = collector.peek()) != null) {
+            for (EventHandler h : handlers) {
+               if (log.isTraceEnabled()) {
+                  log.trace("Handling " + ev + " towards " + h);
+               }
+               try {
+                  Events.dispatch(ev, h);
+               } catch (Exception e) {
+                  log.warn(e.getMessage(), e);
+                  connection.setCondition(new ErrorCondition());
```

Comment:
Its not clear what the intent is here (and in similar snippet further down the method), but I'm not sure this will be having the desired effect.

Setting a brand new ErrorCondition like this will effectively clear any error details for the connection, which are actually the condition/description/info-map details stored inside an ErrorCondition object, which will be empty for the new instance. Setting a condition inside the ErrorCondition doesn't itself lead to it being sent, the endpoint subsequently needs closed before the details are included in the detach/end/close frame then being sent. The connection may not be the right endpoint to set an error, e.g it might be a particular link receiving a delivery event and so that link might be the endpoint to update.